File: src/main/java/io/github/xxfast/design/patterns/creational/builders/shared/Cake.java
package io.github.xxfast.design.patterns.creational.builders.shared;


import org.jetbrains.annotations.Nullable;

import java.util.List;

public class Cake {
  private List<Layers> layers;
  private List<Candle> candles;
  private boolean isGluttonFree;
  private boolean isDairyFree;
  @Nullable private Topper topper;
  @Nullable private Sponge sponge;

  public Cake(
      List<Layers> layers, List<Candle> candles,
      boolean isGluttonFree, boolean isDairyFree,
      @Nullable Topper topper, @Nullable Sponge sponge
  ) {
    this.layers = layers;
    this.candles = candles;
    this.isGluttonFree = isGluttonFree;
    this.isDairyFree = isDairyFree;
    this.topper = topper;
    this.sponge = sponge;
  }
}

File: src/main/kotlin/io/github/xxfast/design/patterns/creational/builder/solution/Cake.kt
package io.github.xxfast.design.patterns.creational.builder.solution

import io.github.xxfast.design.patterns.creational.builders.shared.Layers
import io.github.xxfast.design.patterns.creational.builders.shared.Sponge
import io.github.xxfast.design.patterns.creational.builders.shared.Candle
import io.github.xxfast.design.patterns.creational.builders.shared.Topper

sealed interface CakeRecipe
data class VanillaCakeRecipe(val containsEggs: Boolean) : CakeRecipe

fun Cake(recipe: CakeRecipe) = Cake(
  isDairyFree = recipe is VanillaCakeRecipe && recipe.containsEggs,
)

class Cake(
  val layers: List<Layers> = List(4) { Layers() },
  val toppings: List<Candle> = List(3) { Candle() },
  val isGluttonFree: Boolean = true,
  val isDairyFree: Boolean = false,
  val topper: Topper? = null,
  val sponge: Sponge = Sponge()
)

val car = Cake(isDairyFree = true)